text/0000-const-ub.md

```diff
@@ -0,0 +1,111 @@
+- Feature Name: `const_ub`
+- Start Date: 2020-10-10
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Define UB during const evaluation to lead to an unspecified result for the affected CTFE query, but not otherwise infect the compilation process.
+
+# Motivation
+[motivation]: #motivation
+
+So far, nothing is specified about what happens when `unsafe` code leads to UB during CTFE.
+This is a major blocker for stabilizing `unsafe` operations in const-contexts.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+There are some values that Rust needs to compute at compile-time.
+This includes the initial value of a `const`/`static`, and array lengths (and more general, const generics).
+Computing these initial values is called compile-time function evaluation (CTFE).
+CTFE in Rust is very powerful and permits running almost arbitrary Rust code.
+This raises the question, what happens when there is `unsafe` code and it causes [Undefined Behavior (UB)][UB]?
+
+The answer is that in this case, the final value that is currently being executed is arbitrary.
+For example, when UB arises while computing an array length, then the final array length can be any `usize`, or it can be (partially) uninitialized.
+No guarantees are made about this final value, and it can be different depending on host and target architecture, compiler flags, and more.
+However, UB will not otherwise adversely affect the currently running compiler; type-checking and lints and everything else will work correctly given whatever the result of the CTFE computation is.
```

Comment:
I think it should be noted that the behaviour is also deterministic, because if it were not, that would be adversely affecting the compilation.

Reply:
Yes. The point I was trying to convey but failed is: You don't see two different values of the same constant within a single compilation. Like two crates A and B evaluating a constant defined in C will give the same result, so that D (which depends on A and B) does not end up with a broken type system.